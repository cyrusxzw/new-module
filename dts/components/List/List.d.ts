import React from 'react';
import type { ListProps } from './List.types';
declare const List: React.ForwardRefExoticComponent<
  ListProps & React.RefAttributes<HTMLUListElement>
>;
export { List };
