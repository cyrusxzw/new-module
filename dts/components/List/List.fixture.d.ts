import type { ListProps } from './List.types';
declare const ListFixture: ListProps;
export { ListFixture };
