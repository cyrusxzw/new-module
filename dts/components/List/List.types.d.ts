/// <reference types="react" />
import type { Themes } from '~/types';
declare type ListItem = {
  content: React.ReactNode;
  id: string;
};
declare type ListProps = {
  className?: string;
  items: ListItem[];
  listItemClassName?: string;
  theme?: Themes;
};
export type { ListProps };
