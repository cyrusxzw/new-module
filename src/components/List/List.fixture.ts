import type { ListProps } from './List.types';

const ListFixture: ListProps = {
  items: [
    { content: 'Lorem ipsum dolor sit amet', id: 'Lorem ipsum dolor sit amet' },
    {
      content: 'Mauris interdum odio at mauris pulvinar',
      id: 'Mauris interdum odio at mauris pulvinar',
    },
    {
      content: 'Phasellus tempus consequat erat',
      id: 'Phasellus tempus consequat erat',
    },
    { content: 'Maecenas elementum', id: 'Maecenas elementum' },
    { content: 'Nullam et felis sem', id: 'Nullam et felis sem' },
  ],
};

export { ListFixture };
