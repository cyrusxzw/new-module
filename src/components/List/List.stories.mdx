import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { List } from './List';
import { ListFixture } from './List.fixture';

<Meta title="Components/List" component={List} />

# List

<Canvas>
  <Story name="Base component">
    {(() => {
      const items = knobs.object('items', ListFixture.items, 'Content');
      const theme = knobs.select(
        'theme',
        ['dark', 'light'],
        'dark',
        'Presentation',
      );
      return <List items={items} theme={theme} />;
    })()}
  </Story>
</Canvas>

## API

<ArgsTable of={List} />
