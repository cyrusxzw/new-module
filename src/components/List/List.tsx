import React, { forwardRef } from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import type { Props } from './List.types';
import styles from './List.module.css';

const List = forwardRef<HTMLUListElement, Props>(function ListRef(
  { className, items, listItemClassName, theme },
  ref,
) {
  const currentTheme = useThemeContext(theme, 'dark');
  const classSet = cx(styles.base, styles[currentTheme], className);
  const listItemClassSet = cx(styles.item, listItemClassName);

  return (
    <ul className={classSet} ref={ref}>
      {items.map(({ content, id }) => (
        <li className={listItemClassSet} key={id}>
          {content}
        </li>
      ))}
    </ul>
  );
});

export { List };
