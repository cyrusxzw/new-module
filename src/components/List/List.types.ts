import type { Themes } from '~/types';

type Item = {
  content: React.ReactNode;
  id: string;
};

type Props = {
  className?: string;
  items: Item[];
  listItemClassName?: string;
  theme?: Themes;
};

export type { Props };
