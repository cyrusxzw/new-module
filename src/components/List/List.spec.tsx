import React from 'react';
import { render } from '@testing-library/react';
import { List } from './List';
import { ListFixture } from './List.fixture';

describe('<List />', () => {
  it('should be defined', () => {
    expect(List).toBeDefined();
  });

  it('should render base component correctly', () => {
    const { container } = render(
      <List items={ListFixture.items} theme="dark" />,
    );

    expect(container).toMatchSnapshot();
  });

  it.todo('should have the correct number of items');
});
