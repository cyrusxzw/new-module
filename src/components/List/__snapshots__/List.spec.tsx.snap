// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should render base component correctly 1`] = `
<div>
  <ul
    class="base dark"
  >
    <li
      class="item"
    >
      Lorem ipsum dolor sit amet
    </li>
    <li
      class="item"
    >
      Mauris interdum odio at mauris pulvinar
    </li>
    <li
      class="item"
    >
      Phasellus tempus consequat erat
    </li>
    <li
      class="item"
    >
      Maecenas elementum
    </li>
    <li
      class="item"
    >
      Nullam et felis sem
    </li>
  </ul>
</div>
`;
